polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -173,42 +173,54 @@ class LoopGuardBuilder {
 public:
   static std::unique_ptr<LoopGuardBuilder> create(LoopLikeOpInterface loop);
 
-  LoopGuardBuilder(LoopLikeOpInterface loop) : builder(loop){};
+  LoopGuardBuilder(LoopLikeOpInterface loop) : builder(loop), loop(loop){};
   LoopGuardBuilder(const LoopGuardBuilder &) = delete;
   LoopGuardBuilder(LoopGuardBuilder &&) = delete;
   void operator=(const LoopGuardBuilder &) = delete;
   void operator=(LoopGuardBuilder &&) = delete;
   virtual ~LoopGuardBuilder() = default;
 
-  virtual void guardLoop() const = 0;
+  void guardLoop() const;
 
 protected:
+  static Block *getThenBlock(RegionBranchOpInterface ifOp) {
+    return &*ifOp->getRegion(0).begin();
+  }
+  static Block *getElseBlock(RegionBranchOpInterface ifOp) {
+    return &*ifOp->getRegion(1).begin();
+  }
+  virtual Operation::operand_range getInitVals() const = 0;
+
   mutable OpBuilder builder;
+
+private:
+  void replaceUsesOfLoopReturnValues(RegionBranchOpInterface &) const;
+  virtual RegionBranchOpInterface createGuard() const = 0;
+
+  mutable LoopLikeOpInterface loop;
 };
 
 class SCFLoopGuardBuilder : public LoopGuardBuilder {
 public:
-  SCFLoopGuardBuilder(LoopLikeOpInterface loop) : LoopGuardBuilder(loop) {}
-
-protected:
-  scf::IfOp guard(LoopLikeOpInterface loop) const;
+  SCFLoopGuardBuilder(LoopLikeOpInterface loop)
+      : LoopGuardBuilder(loop), loop(loop) {}
 
 private:
+  RegionBranchOpInterface createGuard() const final;
   virtual Value createGuardExpr() const = 0;
-  virtual scf::IfOp createGuard() const = 0;
-  virtual void replaceUsesOfLoopReturnValues(scf::IfOp &) const {}
+
+  mutable LoopLikeOpInterface loop;
 };
 
 class SCFForGuardBuilder : public SCFLoopGuardBuilder {
 public:
   SCFForGuardBuilder(scf::ForOp loop) : SCFLoopGuardBuilder(loop), loop(loop) {}
 
-  void guardLoop() const final;
-
 private:
   Value createGuardExpr() const final;
-  scf::IfOp createGuard() const final;
-  void replaceUsesOfLoopReturnValues(scf::IfOp &) const final;
+  Operation::operand_range getInitVals() const final {
+    return loop.getInitArgs();
+  }
 
   mutable scf::ForOp loop;
```

Comment:
Now the base class has this member (although of type LoopLikeOpInterface). Can we remove this one then ? It doesn't make sense to me that both the base class and the derived classes duplicate this member, so either the base class has it or the derived classes should have it but not both.

Reply:
done.